### PR TITLE
move render method to webpack's config

### PR DESCRIPTION
**Pull request by garbas**

### CODE REVIEW

**Comment by eliperelman** [comment]
A lot of this just seems redundant. Couldn't we just eliminate `main`/`entry`, and change the default export of `app.js` to this?

```js
export default ({ reducers: {}, initialState: {}, routes: [], Layout, rootElement: document.getElementById('root') }) => {
  const frozen = Immutable.fromJS(initialState);
  const routing = (state = frozen, action) => {
    return action.type === LOCATION_CHANGE ?
      state.merge({ locationBeforeTransitions: action.payload }) :
      state;
  };
  const store = createStore(
    combineReducers({ ...reducers, routing }),
    frozen,
    devtoolsStore
  );
  const history = syncHistoryWithStore(browserHistory, store, {
    selectLocationState: state => state.get('routing').toJS()
  });

  const LayoutWrapper = (props) => (
    <div id="wrapper">
      <Layout {...props} />
      <DevTools />
    </div>
  );

  ReactDOM.render(
    <Provider store={store}>
      <Router history={history}>
        <Route component={LayoutWrapper}>
          {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
        </Route>
      </Router>
    </Provider>,
    rootElement
  );
};
```

**Comment by garbas** [comment]
i know current solution looks hacky, but i think we benefit more, i'll explain why i separated create/render in the first place.

the only part of ``app.js`` that does some IO is actually that render function. it is important to not mix pure function and IO. if we need to use ``store`` or ``history`` in our tests we would have to mock ReactDOM.render. the usual rule of thumb i follow is that you have to push all IO stuff at edges of your application.

but you are actually right, we should return whole Provider eg.
```javascript
export default ({ reducers: {}, initialState: {}, routes: [], Layout, rootElement: document.getElementById('root') }) => {
  const frozen = Immutable.fromJS(initialState);
  const routing = (state = frozen, action) => {
    return action.type === LOCATION_CHANGE ?
      state.merge({ locationBeforeTransitions: action.payload }) :
      state;
  };
  const store = createStore(
    combineReducers({ ...reducers, routing }),
    frozen,
    devtoolsStore
  );
  const history = syncHistoryWithStore(browserHistory, store, {
    selectLocationState: state => state.get('routing').toJS()
  });

  const LayoutWrapper = (props) => (
    <div id="wrapper">
      <Layout {...props} />
      <DevTools />
    </div>
  );

  const App = (
    <Provider store={store}>
      <Router history={history}>
        <Route component={LayoutWrapper}>
          {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
        </Route>
      </Router>
    </Provider>
  );

  return {App, LayoutWrapper, store, history};
};
```

and then the generated code would look like

```javascript
const TMP_SRC_FILE = tmp.fileSync({prefix: 'mozilla-neo-'}).name;
fs.writeFileSync(TMP_SRC_FILE, "require('ReactDOM').render(require('" + TMP_SRC + "').App, document.getElementById('" + PACKAGE.config.root + '"")"
```

**Comment by garbas** [comment]
it might be useful to post few links where i learned all this (maybe i got it all wrong).

apart from learning this first in haskell, where you put all your IO related work at the top level function there i got across this nice explanation of so called *clean architecture* with nice presentation [The  Clean Architecture in Python](https://www.youtube.com/watch?v=DJtef410XaM), slides can be followed up [here[(http://rhodesmill.org/brandon/slides/2014-07-pyohio/clean-architecture). There is also a [Javascript version](http://codepen.io/njmcode/pen/zHyne) of the same example that brandon uses in his talk.

or if you have time read http://haskellbook.com. which is actually worth every money even if you dont plan to use haskell. but i warn you in advance, you will want to use ghcjs/elm/purescript at some point in the future :)

**Comment by eliperelman** [comment]
You could still export the `App` and use it from index while not having the hackiness.

```js
// index.js
ReactDOM.render(App({ reducers, initialState, Layout, routes }));
```

**Comment by garbas** [comment]
ofcourse you can do anything in index.js, but should you? if you don't have
too?

With above ReactDOM.render you just made it hard to test App. You're
basically mixing pure and code with side effects.

I know implementation wise it looks like a hack, I admit that, but from a
person who would have to use neo, it looks much nicer since now you can now
write tests for against App. if we leave ReactDOM.render you would have to

For the same reason Elm has one function at the top level. The rest is
boilerplate which they hide in the framework.

**Comment by eliperelman** [comment]
> With above ReactDOM.render you just made it hard to test App. You're basically mixing pure and code with side effects.

No, `App` would then be side-effect free because it does no rendering to the DOM, it is shifted to the outside caller. Also, `App` is still testable as you can mock `reducers`, `initialState`, `Layout`, and `routes` and test that `App` returns something that you can render with Enzyme. If you bake in the render you introduce side-effects into the `App` function.

Like I posited before, you can also return a function that handles the rendering:

```js
export default ({ reducers: {}, initialState: {}, routes: [], Layout, rootElement: document.getElementById('root') }) => {
  // ...
  const App = (
    <Provider store={store}>
      <Router history={history}>
        <Route component={LayoutWrapper}>
          {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
        </Route>
      </Router>
    </Provider>
  );

  return () => ReactDOM.render(App);
};
```

```js
let render = App({ reducers, initialState, Layout, routes });

render();
```

**Comment by garbas** [comment]
you're right, but i think you forgot one thing, ``src/index.js`` can not be imported from tests, since that would trigger rendering, right? one solution to this is to split create (yet) another file (eg: ``src/app.js``) which exports App and then every neo project will have to do 

``` javascript
import App from './index';
App.render() // or App() whatever calls the render
```

now why would be like to have this boilerplate code around? and it is going the be the same in every project (if i'm not mistaken)?

**Comment by eliperelman** [comment]
Apps shouldn't be testing external dependencies anyway. Also, `src/index.js` should not be imported anywhere, it is the top level execution which triggers the side effects of rendering. There is nothing really testable in `src/index.js`.

**Comment by garbas** [comment]
by importing App you are not testing external dependencies, but this is the only way to get access to ``store`` and ``history`` which you need (sometimes) to test that your component works nicely with hooks in store/history.

File: commands/build/index.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ module.exports = (args, done) => {
     require(path.resolve(process.cwd(), args.options.config)) :
     prodConfig;
 
+  process.env.NODE_ENV = 'production';
```
Comment (eliperelman) [inline commands/build/index.js]: Manually setting the environments is not necessary, as these are set by the cli interface. See:

https://github.com/taskcluster/neo/blob/master/package.json#L47
https://github.com/taskcluster/neo/blob/master/bin/neo#L33

File: commands/init/templates/src/index.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { create, render } from 'mozilla-neo';
+import Appfrom 'mozilla-neo';
```
Comment (eliperelman) [inline commands/init/templates/src/index.js]: Missing space, should be `import App from 'mozilla-neo';`.

File: commands/init/templates/src/index.js [hunk 3]
```diff
@@ -9,8 +9,9 @@ const initialState = {
   title: '<%= data.appTitle %>'
 };
 
-render(
-  create(reducers, initialState),
-  Layout,
-  routes
-);
+export const app = App({reducers, initialState, Layout, routes});
+export default app;
```
Comment (eliperelman) [inline commands/init/templates/src/index.js]: This is redundant. You could just `export default App(...)`.

Comment (eliperelman) [inline commands/init/templates/src/index.js]: Also, space are missing in object  properties:

```js
App({ reducers, initialState, Layout, routes });
```

File: commands/init/templates/src/index.js [hunk 4]
```diff
@@ -9,8 +9,9 @@ const initialState = {
...
+
+if (process.env.NODE_ENV === "testing") {
+  app.render()
```
Comment (eliperelman) [inline commands/init/templates/src/index.js]: Missing semicolon.

File: commands/start/index.js [hunk 5]
```diff
@@ -17,7 +17,11 @@ module.exports = (args, done) => {
   let compiler = webpack(config);
   let server = new DevServer(compiler, config.devServer);
 
-  server.listen(port);
+  server.listen(port, 'localhost', () => {
```
Comment (eliperelman) [inline commands/start/index.js]: The arrow function could probably just be one-lined. Also, use a template string:

```js
server.listen(port, 'localhost', () => console.log(`Listening on http://localhost:${port}`);
```

File: commands/start/index.js [hunk 6]
```diff
@@ -17,7 +17,11 @@ module.exports = (args, done) => {
...
+    console.log('Listening on http://localhost:' + port )
+  });
+
+  process.env.NODE_ENV = 'development';
```
Comment (eliperelman) [inline commands/start/index.js]: Not needed.

File: commands/test/index.js [hunk 7]
```diff
@@ -2,6 +2,7 @@ const path = require('path');
 const { Server } = require('karma');
 
 module.exports = (args) => {
+  process.env.NODE_ENV = 'testing';
```
Comment (eliperelman) [inline commands/test/index.js]: Not needed.

File: config/webpack.dev.js [hunk 8]
```diff
@@ -21,10 +21,18 @@ module.exports = merge(core, {
...
+
+    // Enable history API fallback so HTML5 History API based
+    // routing works. This is a good default that will come
+    // in handy in more complicated setups.
+    historyApiFallback: true,
+
     hot: true,
     progress: true,
-    contentBase: path.join(process.cwd(), 'src')
+    inline: true,
```
Comment (eliperelman) [inline config/webpack.dev.js]: The `inline` property cannot be set from the Node.js API, only from the CLI interface, so we don't use this.

File: config/webpack.core.js [hunk 9]
```diff
@@ -11,7 +11,8 @@ const CWD_NODE_MODULES = path.join(CWD, 'node_modules');
 const ENV = process.env.NODE_ENV;
 const NODE_MODULES = path.join(__dirname, '../node_modules');
 const PACKAGE = require(path.join(CWD, 'package.json'));
-const SRC = path.join(CWD, 'src');
+const SRC_FILE = path.join(CWD, PACKAGE.main || 'src/index.js').slice(0, -3);
```
Comment (eliperelman) [inline config/webpack.core.js]: I'd probably rather see this be:

```js
const SRC_FILE = path.join(CWD, path.basename(PACKAGE.main || 'src/index.js, '.js'));
```

File: src/app.js [hunk 10]
```diff
@@ -32,41 +32,46 @@ let devtoolsStore = IS_PROD ? undefined : compose(
   persistState(location.href.match(/[?&]debug_session=([^&]+)\b/))
 );
 
-export const create = (reducers = {}, initialState = {}) => {
+export default ({ reducers={}, initialState={}, Layout=null, routes=[] }) => {
```
Comment (eliperelman) [inline src/app.js]: Use spaces before and after assignment:

```js
export default ({ reducers = {}, initialState = {}, routes = [], Layout }) => {
```

File: src/app.js [hunk 11]
```diff
@@ -32,41 +32,46 @@ let devtoolsStore = IS_PROD ? undefined : compose(
...
+
   const store = createStore(
     combineReducers({ ...reducers, routing }),
     frozen,
     devtoolsStore
   );
+
   const history = syncHistoryWithStore(browserHistory, store, {
-    selectLocationState: state => state.get('routing').toJS()
+    selectLocationState: state => state.has('routing') ? state.get('routing').toJS() : undefined
```
Comment (eliperelman) [inline src/app.js]: What does the `.get()` do if the state doesn't have routing? Does it return `undefined`?

File: src/app.js [hunk 12]
```diff
@@ -32,41 +32,46 @@ let devtoolsStore = IS_PROD ? undefined : compose(
...
-    <Provider store={app.store}>
-      <Router history={app.history}>
-        <Route component={LayoutWrapper}>
-          {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
-        </Route>
-      </Router>
-    </Provider>,
-    rootElement
-  );
+  App.render = (rootElement=null) => {
```
Comment (eliperelman) [inline src/app.js]: Could probably be written as:

```js
return {
  store,
  history,
  render(rootElement = document.getElementById('root')) {
    ReactDOM.render(
      <Provider store={store}>
         <Router history={history}>
           <Route component={LayoutWrapper}>
             {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
           </Route>
         </Router>
       </Provider>,
       rootElement
     );
  }
};
```

File: commands/build/index.js (hunk 1 above)
Comment (garbas) [inline commands/build/index.js]: got it! i thought we are not handling this.

File: src/app.js (hunk 11 above)
Comment (garbas) [inline src/app.js]: yes. and then it fails when we call ``toJS``.

File: src/app.js (hunk 12 above)
Comment (garbas) [inline src/app.js]: is `rootElement = document.getElementById('root')`` assigned at the call or when it is interpreted?

File: src/app.js (hunk 11 above)
Comment (eliperelman) [inline src/app.js]: OK cool. Just one change then, let's return `null`.

File: src/app.js (hunk 12 above)
Comment (eliperelman) [inline src/app.js]: Call.

**Comment by garbas** [comment]
@eliperelman fixes pushed please another review

i've moved ``app.render()`` call to ``src/run.js`` which ``package.json`` is pointing from ``config.entry`` and ``main`` is pointing to application.

File: commands/init/templates/src/index.js [hunk 13]
```diff
@@ -1,16 +1,13 @@
...
+export const initialState = {
   title: '<%= data.appTitle %>'
 };
 
-render(
-  create(reducers, initialState),
-  Layout,
-  routes
-);
+const default App({ reducers, initialState, Layout, routes });
```
Comment (eliperelman) [inline commands/init/templates/src/index.js]: This isn't exported, it should be, right?

File: src/app.js [hunk 14]
```diff
@@ -32,41 +32,38 @@ let devtoolsStore = IS_PROD ? undefined : compose(
...
-    </Provider>,
-    rootElement
-  );
+  return {
+    store,
+    history,
+    render(rootElement = document.getElementById('root')) {
+      ReactDOM.render(
+        <Provider store={store}>
+           <Router history={history}>
```
Comment (eliperelman) [inline src/app.js]: I think this indentation is off by a space.

**Comment by eliperelman** [comment]
Other than a couple nits, I think this is almost ready. My contention is still how this is rendered. I think we should rename the `src/index.js` file to `src/init.js`. Then we can rename `src/run.js` to `src/index.js`. This would allow us to remove the `config.entry` prop and resort to using `PACKAGE.main` in webpack. This still keeps creation of the app separate from rendering.

File: commands/init/templates/src/index.js (hunk 13 above)
Comment (garbas) [inline commands/init/templates/src/index.js]: true, i'll export it.

File: src/app.js (hunk 14 above)
Comment (garbas) [inline src/app.js]: :) i need to get my vim plugins for javascript in order

**Comment by garbas** [comment]
as i see it "main" is official package.json location to point default package to be loaded. actually main should point to compiled (with babel) version of this app. we shouldn't bend node's rules here plus i can see this being useful in the future since i'd like to have neo's app to be composable. in similar way how elm apps are composable.

the custom thing here is entry point to webpack, which is different then what which i'd like to keep in separate script, to be able to export app (from src/index.js). also this way you can easily avoid importing src/run.js and not worry about side effects.

File: src/app.js [hunk 15]
```diff
@@ -32,41 +32,45 @@ let devtoolsStore = IS_PROD ? undefined : compose(
...
+    history,
+    render(rootElement = document.getElementById('root')) {
+      ReactDOM.render(
+        <Provider store={store}>
+          <Router history={history}>
+            <Route component={LayoutWrapper}>
+              {routes.map(route => <Route key={route.path} path={route.path} component={route.component} />)}
+            </Route>
+          </Router>
+         </Provider>,
```
Comment (eliperelman) [inline src/app.js]: Lines 71-73 need to be shifted to the left one space.

File: commands/start/index.js [hunk 16]
```diff
@@ -17,7 +17,8 @@ module.exports = (args, done) => {
   let compiler = webpack(config);
   let server = new DevServer(compiler, config.devServer);
 
-  server.listen(port);
+  server.listen(port, 'localhost', () => console.log(`Listening on http:\/\/localhost:${port}`));
```
Comment (eliperelman) [inline commands/start/index.js]: Backslashes are unnecessary.

File: src/app.js [hunk 17]
```diff
@@ -32,41 +32,45 @@ let devtoolsStore = IS_PROD ? undefined : compose(
   persistState(location.href.match(/[?&]debug_session=([^&]+)\b/))
 );
 
-export const create = (reducers = {}, initialState = {}) => {
+export default ({ reducers = {}, initialState={}, routes = [], Layout = NOOP }) => {
```
Comment (eliperelman) [inline src/app.js]: Needs spaces: `initialState={}` => `initialState = {}`.

File: commands/init/templates/src/index.js [hunk 18]
```diff
@@ -1,16 +1,13 @@
...
+export const initialState = {
   title: '<%= data.appTitle %>'
 };
 
-render(
-  create(reducers, initialState),
-  Layout,
-  routes
-);
+export const default App({ reducers, initialState, Layout, routes });
```
Comment (eliperelman) [inline commands/init/templates/src/index.js]: This should just be `export default App(...)`.

**Pull request merged**